Golem/Golem.cs

```diff
@@ -451,5 +479,88 @@ public List<String> LogFiles()
         {
             return Provider.ProviderProcess?.Id;
         }
+
+        private async Task SetupSuspendHandlerAsync()
+        {
+            await SuspendHandlerLock.WaitAsync();
+            try {
+                if (_isSuspendHandlerSetUp) {
+                    return;
+                }
+    #pragma warning disable CA1416 // Validate platform compatibility ensured by #if
+                SetupSuspendHandler_Impl();
+    #pragma warning restore CA1416 // Validate platform compatibility
+                _isSuspendHandlerSetUp = true;
+            } finally {
+                SuspendHandlerLock.Release();
+            }
+
+        }
+        private async Task DetachSuspendHandler()
+        {
+            await SuspendHandlerLock.WaitAsync();
+            try {
+#pragma warning disable CA1416 // Validate platform compatibility ensured by #if
+            DetachSuspendHandler_Impl();
+#pragma warning restore CA1416 // Validate platform compatibility
+            _isSuspendHandlerSetUp = false;
+            } finally {
+                SuspendHandlerLock.Release();
+            }
+        }
+
+        #if OS_WINDOWS
+
+        [SupportedOSPlatform("Windows")]
+        private void SetupSuspendHandler_Impl()
+        {
+            _logger.LogInformation("Setting up Suspend handler (Windows).");
+            SystemEvents.PowerModeChanged += OnPowerModeChanged;
+        }
+
+        [SupportedOSPlatform("Windows")]
+        private void DetachSuspendHandler_Impl()
+        {
+            _logger.LogInformation("Detaching Suspend handler (Windows).");
+            SystemEvents.PowerModeChanged -= OnPowerModeChanged;
+        }
+
+        [SupportedOSPlatform("Windows")]
+        private void OnPowerModeChanged(object sender, PowerModeChangedEventArgs e)
+        {
+            // Start and Stop tasks will be won't be waited upon in this event handler.
+            switch (e.Mode) {
+                case PowerModes.Suspend:
+                    _logger.LogInformation("Received Suspend event - stopping Golem.");
+                    InternalStop().ContinueWith(t => Console.WriteLine(t.Exception), TaskContinuationOptions.OnlyOnFaulted);
+                    break;
+                case PowerModes.Resume:
+                    _logger.LogInformation("Received Resume event - starting Golem.");
+                    RestartOnResume().ContinueWith(t => Console.WriteLine(t.Exception), TaskContinuationOptions.OnlyOnFaulted);
+                    break;
+            }
+        }
+
+        private async Task RestartOnResume()
+        {
+            // We are trying to stop Golem when the system is Resumed from a sleep
+            // because we cannot ensure that Suspend handler will execute to completion
+            // before the system is suspended.
+            await InternalStop();
```

Comment:
sanity check: note that it may result in calling `InternalStop()` twice - once by the handler and once by us here. I looked at the definition of `InternalStop()` and `Stop()` methods it calls and it looks safe to me, but also please take another look.